### DISCUSSION TITLE
Allow Actions to be pulled from any registry

### DISCUSSION THREAD

**Discussion by jacobweinstock** [tinkerbell/tink#586] [Ideas]
Allowing [Actions](https://docs.tinkerbell.org/actions/action-architecture/) to be pulled from any registry has been brought up many times in [community meetings](https://docs.google.com/document/d/1cEObfvQ9Tdp8zIIIg9O7P5i3CKaSj2t3JTxEufDxwWs/edit?_hsmi=95274993&_hsenc=p2ANqtz-_hOk14E6kqohN-mM9-b6FwJ5ZG98yoT-iydtVrQEjMAEyJaGDzY4Buu5zekT1yE-rdUn7yVTvkS6vJyCv5oyVAu6GZlA), in [slack](https://cloud-native.slack.com/archives/C01SRB41GMT), and through GitHub issues (https://github.com/tinkerbell/tink/issues/469, https://github.com/tinkerbell/tink/issues/303, https://github.com/tinkerbell/sandbox/issues/45). I would like to document, discuss, and capture the context and touchpoints for this change. 

CC @abhinavmpandey08

# Context

## Current State
Currently, only a single container registry can be used in the Tinkerbell stack (for tink-worker and Actions image downloading). This registry must-have auth enabled (I believe this to be the case, but I'm only about 85% sure) as well. 

## Desired State
Any combination of registries, with or without auth, is supported in Actions and for the tink-worker download.

# Touchpoints in the stack

## Tink worker

[Tink worker](https://github.com/tinkerbell/tink/tree/main/cmd/tink-worker) is what ultimately crafts the registry's URL and then performs the pulling down of a container image. A single registry is passed to the tink-worker at its runtime. See [here](https://github.com/tinkerbell/tink/blob/41bf3415eb244609cb9d35548b16f26f26080efa/cmd/tink-worker/cmd/root.go#L50)). Then that registry is prepended to all image names (image names defined in each action). See [here](https://github.com/tinkerbell/tink/blob/41bf3415eb244609cb9d35548b16f26f26080efa/cmd/tink-worker/internal/registry.go#L75). 

## Boots

[Boots](https://github.com/tinkerbell/boots), at runtime, requires the following environment variables: [`DOCKER_REGISTRY`](https://github.com/tinkerbell/boots/blob/ba3a3fef424ebfd7125b08ae99dcb9631bc911a8/installers/osie/mirror.go#L55),[`REGISTRY_USERNAME`](https://github.com/tinkerbell/boots/blob/ba3a3fef424ebfd7125b08ae99dcb9631bc911a8/installers/osie/mirror.go#L58), [`REGISTRY_PASSWORD`](https://github.com/tinkerbell/boots/blob/ba3a3fef424ebfd7125b08ae99dcb9631bc911a8/installers/osie/mirror.go#L59). These variables get passed on to the in-memory operating system ([Hook](https://github.com/tinkerbell/hook)) as [kernel command-line parameters](https://www.kernel.org/doc/html/v4.14/admin-guide/kernel-parameters.html). See [here](https://github.com/tinkerbell/boots/blob/1a561842d4643844c7261173b59957d90c16cfc8/installers/osie/main.go#L106-L111).

## Hook

[Hook](https://github.com/tinkerbell/hook) is responsible for starting up the tink-worker. See [here](https://github.com/tinkerbell/hook/blob/9982f1af8e8ee7600b991c205650f6ea5123749f/bootkit/main.go#L145). As a part of the tink-worker start-up process, Hook parses its kernel command-line parameters in order to get the container registry and auth. See [here](https://github.com/tinkerbell/hook/blob/main/bootkit/main.go#L49-L55). Hook then uses the parsed values as environment variables in the container start-up of tink-worker. See [here](https://github.com/tinkerbell/hook/blob/main/bootkit/main.go#L70-L72).

**Comment by jacobweinstock** [comment]
If anyone knows other touchpoints, please do comment. thank you!

**Comment by abhay-krishna** [comment]
Hey @jacobweinstock, can you cc [Abhinav](https://github.com/abhinavmpandey08) from our team, who is leading the efforts on this? Thanks!

**Reply by jacobweinstock** [reply]
Ah, so sorry about that @abhay-krishna!

CC @abhinavmpandey08

**Reply by abhay-krishna** [reply]
I am not part of the main efforts but can definitely contribute with thoughts and touchpoints.
